VSDiagnostics/VSDiagnostics/VSDiagnostics/Diagnostics/Attributes/OnPropertyChangedWithoutCallerMemberName/OnPropertyChangedWithoutCallerMemberNameAnalyzer.cs

```diff
@@ -0,0 +1,93 @@
+    {
+        private const string DiagnosticId = nameof(OnPropertyChangedWithoutCallerMemberNameAnalyzer);
+        private const DiagnosticSeverity Severity = DiagnosticSeverity.Warning;
+
+        private static readonly string Category = VSDiagnosticsResources.AttributesCategory;
+        private static readonly string Message = VSDiagnosticsResources.OnPropertyChangedWithoutCallerMemberNameAnalyzerMessage;
+        private static readonly string Title = VSDiagnosticsResources.OnPropertyChangedWithoutCallerMemberNameAnalyzerTitle;
+
+        internal static DiagnosticDescriptor Rule => new DiagnosticDescriptor(DiagnosticId, Title, Message, Category, Severity, true);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(AnalyzeSymbol, SyntaxKind.ClassDeclaration);
+        }
+
+        private void AnalyzeSymbol(SyntaxNodeAnalysisContext context)
+        {
+            var classDeclaration = (ClassDeclarationSyntax) context.Node;
+
+            // class must implement INotifyPropertyChanged
+            if (!ClassImplementsINotifyPropertyChanged(context.SemanticModel, classDeclaration))
+            {
+                return;
+            }
+
+            var methods = classDeclaration.DescendantNodes().OfType<MethodDeclarationSyntax>();
+
+            foreach (var method in methods)
+            {
+                var declaredSymbol = context.SemanticModel.GetDeclaredSymbol(method);
+                if (declaredSymbol == null || declaredSymbol.MetadataName != "OnPropertyChanged")
+                {
+                    continue;
+                }
+
+                if (method.ParameterList.Parameters.Count != 1)
+                {
+                    continue;
+                }
+
+                var param = method.ParameterList.Parameters.First();
+                var paramType = method.ParameterList.Parameters.First().Type as PredefinedTypeSyntax;
+
+                if (paramType == null || !paramType.Keyword.IsKind(SyntaxKind.StringKeyword))
+                {
+                    continue;
+                }
+
+                if (param.AttributeLists.Any(a => a.Attributes.Any() && a.Attributes.Any(t =>
+                {
+                    var symbol = context.SemanticModel.GetSymbolInfo(t).Symbol;
+                    return symbol != null && symbol.ContainingSymbol.MetadataName == typeof (CallerMemberNameAttribute).Name;
+                })))
+                {
+                    return;
+                }
+
+                context.ReportDiagnostic(Diagnostic.Create(Rule, method.GetLocation()));
+                return;
+            }
+        }
+
+        private bool ClassImplementsINotifyPropertyChanged(SemanticModel semanticModel, ClassDeclarationSyntax classDeclaration)
```

Comment:
Can we extract this out as an extension method on ClassDeclarationSyntax? You can put it in the Extensions class -- it seems like a useful helper.